File: android/src/main/kotlin/com/example/motion_core/MotionCorePlugin.kt
package com.example.motion_core

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Android implementation of motion_core.
 *
 * Sample payload (a Float64List of [PAYLOAD_LENGTH] doubles). Keep in sync with
 * `lib/motion_core.dart` and `ios/Classes/MotionCorePlugin.swift`:
 *
 * ```
 * [0..3]   attitude quaternion x, y, z, w   (device -> world, East-North-Up)
 * [4..6]   gravity x, y, z                  (m/s², TYPE_GRAVITY convention: flat face-up ≈ +9.81 on Z)
 * [7..9]   user acceleration x, y, z        (m/s², TYPE_LINEAR_ACCELERATION)
 * [10]     heading accuracy                 (radians, -1 when unavailable)
 * [11..13] rotation rate x, y, z            (rad/s, NaN when no gyroscope)
 * [14..16] magnetic field x, y, z           (µT, NaN when unavailable / arbitrary frame)
 * [17]     magnetic field calibration       (-1 uncalibrated, 0 low, 1 medium, 2 high)
 * [18]     effective reference frame index  (see FRAME_* constants)
 * [19]     timestamp                        (seconds since boot)
 * ```
 */
class MotionCorePlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    SensorEventListener {

    companion object {
        const val METHOD_CHANNEL_NAME = "dev.flutter/motion_core_method_channel"
        const val EVENT_CHANNEL_NAME = "dev.flutter/motion_core_event_channel"

        const val PAYLOAD_LENGTH = 20

        /** ~60 Hz, the same default as the iOS implementation. */
        const val DEFAULT_SAMPLING_PERIOD_US = 16_667

        // Values of AttitudeReferenceFrame.index on the Dart side.
        const val FRAME_ARBITRARY_Z_VERTICAL = 0
        const val FRAME_ARBITRARY_CORRECTED_Z_VERTICAL = 1
        const val FRAME_MAGNETIC_NORTH_Z_VERTICAL = 2
        const val FRAME_TRUE_NORTH_Z_VERTICAL = 3

        fun isNorthReferenced(frame: Int): Boolean =
            frame == FRAME_MAGNETIC_NORTH_Z_VERTICAL || frame == FRAME_TRUE_NORTH_Z_VERTICAL

        /**
         * Pure-Kotlin equivalent of [SensorManager.getQuaternionFromVector] that also
         * normalises the result. Returns `[x, y, z, w]`.
         *
         * [length] is the number of meaningful entries in [rotationVector]; when it is
         * 3 the scalar part is reconstructed from the unit-length constraint.
         */
        internal fun quaternionFromRotationVector(rotationVector: FloatArray, length: Int): DoubleArray {
            val x = rotationVector[0].toDouble()
            val y = rotationVector[1].toDouble()
            val z = rotationVector[2].toDouble()
            val w = if (length >= 4) {
                rotationVector[3].toDouble()
            } else {
                val remainder = 1.0 - x * x - y * y - z * z
                if (remainder > 0) sqrt(remainder) else 0.0
            }
            val norm = sqrt(x * x + y * y + z * z + w * w)
            return if (norm > 0 && norm.isFinite()) {
                doubleArrayOf(x / norm, y / norm, z / norm, w / norm)
            } else {
                doubleArrayOf(0.0, 0.0, 0.0, 1.0)
            }
        }

        /** Maps SensorManager.SENSOR_STATUS_* to the iOS CMMagneticFieldCalibrationAccuracy scale. */
        internal fun magneticAccuracyFromSensorStatus(status: Int): Double = when (status) {
            SensorManager.SENSOR_STATUS_ACCURACY_HIGH -> 2.0
            SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM -> 1.0
            SensorManager.SENSOR_STATUS_ACCURACY_LOW -> 0.0
            else -> -1.0
        }

        /** Builds one sample. Pure function so it can be unit tested without Android. */
        internal fun buildPayload(
            rotationVector: FloatArray,
            rotationVectorLength: Int,
            gravity: FloatArray,
            linearAcceleration: FloatArray,
            rotationRate: FloatArray?,
            magneticField: FloatArray?,
            magneticAccuracyStatus: Int,
            frame: Int,
            timestampNanos: Long,
        ): DoubleArray {
            val payload = DoubleArray(PAYLOAD_LENGTH) { Double.NaN }

            val q = quaternionFromRotationVector(rotationVector, rotationVectorLength)
            payload[0] = q[0]
            payload[1] = q[1]
            payload[2] = q[2]
            payload[3] = q[3]

            payload[4] = gravity[0].toDouble()
            payload[5] = gravity[1].toDouble()
            payload[6] = gravity[2].toDouble()

            payload[7] = linearAcceleration[0].toDouble()
            payload[8] = linearAcceleration[1].toDouble()
            payload[9] = linearAcceleration[2].toDouble()

            // values[4] is only defined for the (geomagnetic) rotation vector and is -1
            // when the HAL cannot estimate it.
            payload[10] = if (rotationVectorLength >= 5 && isNorthReferenced(frame)) {
                rotationVector[4].toDouble()
            } else {
                -1.0
            }

            if (rotationRate != null) {
                payload[11] = rotationRate[0].toDouble()
                payload[12] = rotationRate[1].toDouble()
                payload[13] = rotationRate[2].toDouble()
            }

            if (magneticField != null) {
                payload[14] = magneticField[0].toDouble()
                payload[15] = magneticField[1].toDouble()
                payload[16] = magneticField[2].toDouble()
                payload[17] = magneticAccuracyFromSensorStatus(magneticAccuracyStatus)
            }

            payload[18] = frame.toDouble()
            payload[19] = timestampNanos / 1_000_000_000.0
            return payload
        }
    }

    private var sensorManager: SensorManager? = null
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null

    // Lazy so the class can be instantiated in plain JVM unit tests.
    private val mainHandler: Handler by lazy { Handler(Looper.getMainLooper()) }

    // Configuration (mutable through the "configure" method call).
    private var samplingPeriodUs = DEFAULT_SAMPLING_PERIOD_US
    private var requestedFrame = FRAME_MAGNETIC_NORTH_Z_VERTICAL

    // Sensors registered for the current stream.
    private var effectiveFrame = FRAME_MAGNETIC_NORTH_Z_VERTICAL
    private var rotationSensor: Sensor? = null
    private var gravitySensor: Sensor? = null
    private var linearAccelerationSensor: Sensor? = null
    private var gyroscopeSensor: Sensor? = null
    private var magnetometerSensor: Sensor? = null
    private var isStreaming = false

    // Latest samples. Sensor callbacks are delivered on the main thread (see mainHandler).
    private val rotationVector = FloatArray(5)
    private var rotationVectorLength = 0
    private val gravity = FloatArray(3)
    private val linearAcceleration = FloatArray(3)
    private val rotationRate = FloatArray(3)
    private val magneticField = FloatArray(3)
    private var hasGravity = false
    private var hasLinearAcceleration = false
    private var hasRotationRate = false
    private var hasMagneticField = false
    private var magneticAccuracyStatus = SensorManager.SENSOR_STATUS_UNRELIABLE

    // region FlutterPlugin

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        sensorManager =
            binding.applicationContext.getSystemService(Context.SENSOR_SERVICE) as? SensorManager

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL_NAME).also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL_NAME).also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // Make sure sensors are released even if Dart never cancelled the stream
        // (e.g. the engine is destroyed while a subscription is active).
        stopSensors()
        eventSink = null
        methodChannel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        methodChannel = null
        eventChannel = null
        sensorManager = null
    }

    // endregion

    // region MethodChannel

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(isDeviceMotionAvailable())
            "availableReferenceFrames" -> result.success(availableReferenceFrames())
            "configure" -> configure(call, result)
            else -> result.notImplemented()
        }
    }

    private fun configure(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        if (args == null) {
            result.error("INVALID_ARGUMENT", "configure expects a map of options.", null)
            return
        }
        val intervalMicros = (args["updateIntervalMicros"] as? Number)?.toLong()
        val frame = (args["referenceFrame"] as? Number)?.toInt()

        if (intervalMicros != null && intervalMicros <= 0) {
            result.error("INVALID_ARGUMENT", "updateIntervalMicros must be greater than zero.", null)
            return
        }
        if (frame != null && (frame < FRAME_ARBITRARY_Z_VERTICAL || frame > FRAME_TRUE_NORTH_Z_VERTICAL)) {
            result.error("INVALID_ARGUMENT", "referenceFrame must be between 0 and 3.", null)
            return
        }

        intervalMicros?.let { samplingPeriodUs = min(it, Int.MAX_VALUE.toLong()).toInt() }
        frame?.let { requestedFrame = it }
        // showsCalibrationDisplay is iOS-only and intentionally ignored.

        if (isStreaming) {
            stopSensors()
            startSensors()
        }
        result.success(null)
    }

    private fun isDeviceMotionAvailable(): Boolean {
        val sm = sensorManager ?: return false
        val hasRotationVector = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null ||
            sm.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR) != null ||
            sm.getDefaultSensor(Sensor.TYPE_GAME_ROTATION_VECTOR) != null
        return hasRotationVector &&
            sm.getDefaultSensor(Sensor.TYPE_GRAVITY) != null &&
            sm.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION) != null
    }

    private fun availableReferenceFrames(): List<Int> {
        val sm = sensorManager ?: return emptyList()
        val frames = mutableListOf<Int>()
        if (sm.getDefaultSensor(Sensor.TYPE_GAME_ROTATION_VECTOR) != null) {
            frames.add(FRAME_ARBITRARY_Z_VERTICAL)
        }
        if (sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) != null ||
            sm.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR) != null
        ) {
            frames.add(FRAME_MAGNETIC_NORTH_Z_VERTICAL)
        }
        return frames
    }

    // endregion

    // region EventChannel

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (events == null) return
        if (!isDeviceMotionAvailable()) {
            events.error(
                "UNAVAILABLE",
                "Required motion sensors (rotation vector, gravity, linear acceleration) are not available on this device.",
                null,
            )
            events.endOfStream()
            return
        }
        eventSink = events
        startSensors()
    }

    override fun onCancel(arguments: Any?) {
        stopSensors()
        eventSink = null
    }

    // endregion

    // region Sensors

    private fun resolveSensors(sm: SensorManager) {
        val game = sm.getDefaultSensor(Sensor.TYPE_GAME_ROTATION_VECTOR)
        val full = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
        val geomagnetic = sm.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)

        // Candidate (sensor, frame it produces) pairs, most preferred first.
        val northReferenced = listOf(
            full to FRAME_MAGNETIC_NORTH_Z_VERTICAL,
            geomagnetic to FRAME_MAGNETIC_NORTH_Z_VERTICAL,
            game to FRAME_ARBITRARY_Z_VERTICAL,
        )
        val arbitrary = listOf(
            game to FRAME_ARBITRARY_Z_VERTICAL,
            full to FRAME_MAGNETIC_NORTH_Z_VERTICAL,
            geomagnetic to FRAME_MAGNETIC_NORTH_Z_VERTICAL,
        )
        val candidates = if (isNorthReferenced(requestedFrame)) northReferenced else arbitrary
        val chosen = candidates.firstOrNull { it.first != null }

        rotationSensor = chosen?.first
        effectiveFrame = chosen?.second ?: requestedFrame
        gravitySensor = sm.getDefaultSensor(Sensor.TYPE_GRAVITY)
        linearAccelerationSensor = sm.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)
        gyroscopeSensor = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
        // Like Core Motion, only touch the magnetometer for north-referenced frames.
        magnetometerSensor = if (isNorthReferenced(effectiveFrame)) {
            sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
        } else {
            null
        }
    }

    private fun resetSamples() {
        rotationVectorLength = 0
        hasGravity = false
        hasLinearAcceleration = false
        hasRotationRate = false
        hasMagneticField = false
        magneticAccuracyStatus = SensorManager.SENSOR_STATUS_UNRELIABLE
    }

    private fun startSensors() {
        val sm = sensorManager ?: return
        resolveSensors(sm)
        resetSamples()

        val handler = mainHandler
        listOfNotNull(
            rotationSensor,
            gravitySensor,
            linearAccelerationSensor,
            gyroscopeSensor,
            magnetometerSensor,
        ).forEach { sensor ->
            sm.registerListener(this, sensor, samplingPeriodUs, handler)
        }
        isStreaming = true
    }

    private fun stopSensors() {
        if (isStreaming) {
            sensorManager?.unregisterListener(this)
        }
        isStreaming = false
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        if (sensor?.type == Sensor.TYPE_MAGNETIC_FIELD) {
            magneticAccuracyStatus = accuracy
        }
    }

    override fun onSensorChanged(event: SensorEvent?) {
        event ?: return
        val values = event.values ?: return

        when (event.sensor.type) {
            Sensor.TYPE_ROTATION_VECTOR,
            Sensor.TYPE_GAME_ROTATION_VECTOR,
            Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR,
            -> {
                if (event.sensor.type != rotationSensor?.type) return
                val n = min(values.size, rotationVector.size)
                System.arraycopy(values, 0, rotationVector, 0, n)
                rotationVectorLength = n
                if (hasGravity && hasLinearAcceleration) {
                    emitSample(event.timestamp)
                }
            }
            Sensor.TYPE_GRAVITY -> {
                copy3(values, gravity)
                hasGravity = true
            }
            Sensor.TYPE_LINEAR_ACCELERATION -> {
                copy3(values, linearAcceleration)
                hasLinearAcceleration = true
            }
            Sensor.TYPE_GYROSCOPE -> {
                copy3(values, rotationRate)
                hasRotationRate = true
            }
            Sensor.TYPE_MAGNETIC_FIELD -> {
                copy3(values, magneticField)
                magneticAccuracyStatus = event.accuracy
                hasMagneticField = true
            }
        }
    }

    private fun copy3(source: FloatArray, destination: FloatArray) {
        System.arraycopy(source, 0, destination, 0, min(3, source.size))
    }

    private fun emitSample(timestampNanos: Long) {
        val sink = eventSink ?: return
        if (rotationVectorLength < 3) return
        val payload = buildPayload(
            rotationVector = rotationVector,
            rotationVectorLength = rotationVectorLength,
            gravity = gravity,
            linearAcceleration = linearAcceleration,
            rotationRate = if (hasRotationRate) rotationRate else null,
            magneticField = if (hasMagneticField) magneticField else null,
            magneticAccuracyStatus = magneticAccuracyStatus,
            frame = effectiveFrame,
            timestampNanos = timestampNanos,
        )
        sink.success(payload)
    }

    // endregion
}

File: ios/motion_core/Sources/motion_core/MotionCorePlugin.swift
import CoreMotion
import Flutter
import UIKit

/// iOS implementation of motion_core.
///
/// Sample payload (a Float64List of `payloadLength` doubles). Keep in sync with
/// `lib/motion_core.dart` and the Android plugin:
///
/// ```
/// [0..3]   attitude quaternion x, y, z, w   (device -> world, East-North-Up)
/// [4..6]   gravity x, y, z                  (m/s², flat face-up ≈ +9.81 on Z)
/// [7..9]   user acceleration x, y, z        (m/s², positive in the direction of motion)
/// [10]     heading accuracy                 (always -1: Core Motion does not expose it)
/// [11..13] rotation rate x, y, z            (rad/s)
/// [14..16] magnetic field x, y, z           (µT, NaN when unavailable / arbitrary frame)
/// [17]     magnetic field calibration       (-1 uncalibrated, 0 low, 1 medium, 2 high)
/// [18]     effective reference frame index  (see `ReferenceFrame`)
/// [19]     timestamp                        (seconds since boot)
/// ```
///
/// Core Motion reports gravity and user acceleration in G with the opposite sign of
/// Android (a device lying face up reads `(0, 0, -1)`), and its north-referenced frames
/// put +X on north. Both are converted here so Dart sees one convention.
public class MotionCorePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    static let methodChannelName = "dev.flutter/motion_core_method_channel"
    static let eventChannelName = "dev.flutter/motion_core_event_channel"
    static let payloadLength = 20
    static let standardGravity = 9.80665
    /// ~60 Hz, the same default as the Android implementation.
    static let defaultUpdateInterval: TimeInterval = 1.0 / 60.0

    /// Values of `AttitudeReferenceFrame.index` on the Dart side.
    enum ReferenceFrame: Int, CaseIterable {
        case arbitraryZVertical = 0
        case arbitraryCorrectedZVertical = 1
        case magneticNorthZVertical = 2
        case trueNorthZVertical = 3

        var coreMotion: CMAttitudeReferenceFrame {
            switch self {
            case .arbitraryZVertical: return .xArbitraryZVertical
            case .arbitraryCorrectedZVertical: return .xArbitraryCorrectedZVertical
            case .magneticNorthZVertical: return .xMagneticNorthZVertical
            case .trueNorthZVertical: return .xTrueNorthZVertical
            }
        }

        var isAvailable: Bool {
            CMMotionManager.availableAttitudeReferenceFrames().contains(coreMotion)
        }

        /// Fallback order when the requested frame is not supported: stay as close as
        /// possible to the requested semantics.
        var fallbackOrder: [ReferenceFrame] {
            switch self {
            case .trueNorthZVertical:
                return [.trueNorthZVertical, .magneticNorthZVertical, .arbitraryCorrectedZVertical, .arbitraryZVertical]
            case .magneticNorthZVertical:
                return [.magneticNorthZVertical, .arbitraryCorrectedZVertical, .arbitraryZVertical]
            case .arbitraryCorrectedZVertical:
                return [.arbitraryCorrectedZVertical, .arbitraryZVertical, .magneticNorthZVertical]
            case .arbitraryZVertical:
                return [.arbitraryZVertical, .arbitraryCorrectedZVertical, .magneticNorthZVertical]
            }
        }

        var resolved: ReferenceFrame {
            fallbackOrder.first { $0.isAvailable } ?? self
        }
    }

    private let motionManager = CMMotionManager()
    private var eventSink: FlutterEventSink?
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?

    // Configuration (mutable through the "configure" method call).
    private var updateInterval: TimeInterval = MotionCorePlugin.defaultUpdateInterval
    private var requestedFrame: ReferenceFrame = .magneticNorthZVertical
    private var showsCalibrationDisplay = true

    // MARK: - FlutterPlugin

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = MotionCorePlugin()
        let methodChannel = FlutterMethodChannel(
            name: methodChannelName, binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(
            name: eventChannelName, binaryMessenger: registrar.messenger())
        instance.methodChannel = methodChannel
        instance.eventChannel = eventChannel

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        // Publishing is what makes the engine call detachFromEngine(for:) on teardown.
        registrar.publish(instance)
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        // Release Core Motion even if Dart never cancelled the stream.
        stopDeviceMotionUpdates()
        eventSink = nil
        eventChannel?.setStreamHandler(nil)
        methodChannel?.setMethodCallHandler(nil)
        eventChannel = nil
        methodChannel = nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            result(motionManager.isDeviceMotionAvailable)
        case "availableReferenceFrames":
            result(ReferenceFrame.allCases.filter { $0.isAvailable }.map { $0.rawValue })
        case "configure":
            configure(call.arguments, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func configure(_ arguments: Any?, result: FlutterResult) {
        guard let args = arguments as? [String: Any] else {
            result(FlutterError(
                code: "INVALID_ARGUMENT", message: "configure expects a map of options.", details: nil))
            return
        }

        var newInterval = updateInterval
        var newFrame = requestedFrame
        var newShowsCalibrationDisplay = showsCalibrationDisplay

        if let micros = args["updateIntervalMicros"] as? NSNumber {
            let seconds = micros.doubleValue / 1_000_000
            guard seconds > 0 else {
                result(FlutterError(
                    code: "INVALID_ARGUMENT",
                    message: "updateIntervalMicros must be greater than zero.", details: nil))
                return
            }
            newInterval = seconds
        }
        if let rawFrame = args["referenceFrame"] as? NSNumber {
            guard let frame = ReferenceFrame(rawValue: rawFrame.intValue) else {
                result(FlutterError(
                    code: "INVALID_ARGUMENT",
                    message: "referenceFrame must be between 0 and 3.", details: nil))
                return
            }
            newFrame = frame
        }
        if let shows = args["showsCalibrationDisplay"] as? Bool {
            newShowsCalibrationDisplay = shows
        }

        updateInterval = newInterval
        requestedFrame = newFrame
        showsCalibrationDisplay = newShowsCalibrationDisplay

        if motionManager.isDeviceMotionActive {
            stopDeviceMotionUpdates()
            startDeviceMotionUpdates()
        }
        result(nil)
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
        -> FlutterError?
    {
        guard motionManager.isDeviceMotionAvailable else {
            // Deliver the error through the sink so Dart listeners actually receive it,
            // then close the stream.
            events(FlutterError(
                code: "UNAVAILABLE",
                message: "Device motion is not available on this device.", details: nil))
            events(FlutterEndOfEventStream)
            return nil
        }
        eventSink = events
        startDeviceMotionUpdates()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        stopDeviceMotionUpdates()
        eventSink = nil
        return nil
    }

    // MARK: - Core Motion

    private func startDeviceMotionUpdates() {
        let frame = requestedFrame.resolved
        motionManager.deviceMotionUpdateInterval = updateInterval
        motionManager.showsDeviceMovementDisplay = showsCalibrationDisplay

        // Deliver on the main queue: platform channels must be used from the platform thread.
        motionManager.startDeviceMotionUpdates(using: frame.coreMotion, to: .main) {
            [weak self] motion, error in
            guard let self = self, let sink = self.eventSink else { return }

            if let error = error as NSError? {
                // Core Motion reports this while it waits for the user to move the device
                // so the compass can calibrate; samples resume on their own afterwards.
                let requiresMovement = error.domain == CMErrorDomain
                    && error.code == Int(CMErrorDeviceRequiresMovement.rawValue)
                if !requiresMovement {
                    sink(FlutterError(
                        code: "MOTION_ERROR", message: error.localizedDescription,
                        details: error.domain + "/" + String(error.code)))
                }
                return
            }
            guard let motion = motion else { return }
            sink(FlutterStandardTypedData(float64: MotionCorePlugin.encode(motion, frame: frame)))
        }
    }

    private func stopDeviceMotionUpdates() {
        if motionManager.isDeviceMotionActive {
            motionManager.stopDeviceMotionUpdates()
        }
    }

    /// Converts a `CMDeviceMotion` sample into the shared payload layout.
    static func encode(_ motion: CMDeviceMotion, frame: ReferenceFrame) -> Data {
        var payload = [Double](repeating: .nan, count: payloadLength)

        // Core Motion's north-referenced frames are X-north / Y-west / Z-up. Rotate the
        // world frame by +90° about Z so it becomes East-North-Up like Android:
        // q_enu = q_z(90°) ⊗ q_ios. For the arbitrary frames this only changes the
        // (already arbitrary) yaw origin, so it is applied uniformly.
        let q = motion.attitude.quaternion
        let c = (0.5).squareRoot()  // cos(45°) == sin(45°)
        payload[0] = c * (q.x - q.y)
        payload[1] = c * (q.x + q.y)
        payload[2] = c * (q.z + q.w)
        payload[3] = c * (q.w - q.z)

        // G -> m/s², and flip the sign to the Android / sensors_plus convention.
        let scale = -standardGravity
        payload[4] = motion.gravity.x * scale
        payload[5] = motion.gravity.y * scale
        payload[6] = motion.gravity.z * scale
        payload[7] = motion.userAcceleration.x * scale
        payload[8] = motion.userAcceleration.y * scale
        payload[9] = motion.userAcceleration.z * scale

        payload[10] = -1  // Heading accuracy is not exposed by Core Motion.

        payload[11] = motion.rotationRate.x
        payload[12] = motion.rotationRate.y
        payload[13] = motion.rotationRate.z

        // With the arbitrary frames Core Motion leaves the field at zero / uncalibrated.
        let magnetic = motion.magneticField
        let hasField = magnetic.accuracy != .uncalibrated
            || magnetic.field.x != 0 || magnetic.field.y != 0 || magnetic.field.z != 0
        if hasField {
            payload[14] = magnetic.field.x
            payload[15] = magnetic.field.y
            payload[16] = magnetic.field.z
            payload[17] = Double(magnetic.accuracy.rawValue)
        }

        payload[18] = Double(frame.rawValue)
        payload[19] = motion.timestamp

        return payload.withUnsafeBufferPointer { Data(buffer: $0) }
    }
}
